pkg/lib/messaging/sender.go

```diff
@@ -249,26 +248,16 @@ func (s *Sender) SendWhatsappImmediately(ctx context.Context, msgType translatio
 	return nil
 }
 
-func (s *Sender) testModeSendWhatsapp(opts *whatsapp.SendTemplateOptions) {
-	data, _ := json.MarshalIndent(opts.Components, "", "  ")
-
+func (s *Sender) testModeSendWhatsapp(opts *whatsapp.SendAuthenticationOTPOptions) {
 	s.Logger.
 		WithField("recipient", opts.To).
-		WithField("template_name", opts.TemplateName).
-		WithField("language", opts.Language).
-		WithField("components", string(data)).
-		WithField("namespace", opts.Namespace).
+		WithField("otp", opts.OTP).
 		Warn("sending whatsapp is suppressed in test mode")
 }
 
-func (s *Sender) devModeSendWhatsapp(opts *whatsapp.SendTemplateOptions) {
-	data, _ := json.MarshalIndent(opts.Components, "", "  ")
-
+func (s *Sender) devModeSendWhatsapp(opts *whatsapp.SendAuthenticationOTPOptions) {
 	s.Logger.
 		WithField("recipient", opts.To).
-		WithField("template_name", opts.TemplateName).
-		WithField("language", opts.Language).
-		WithField("components", string(data)).
-		WithField("namespace", opts.Namespace).
```

Comment:
I guess we want to log language, template name, namespace here, because it could be useful when debugging whatsapp otp.

Reply:
Those information is only available in the whatsapp/Service. In messaging/Sender, the available information is just the information contained in whatsapp/SendAuthenticationOTPOptions.

The previous behavior is a bit hacky. If dev mode is enabled, we synthesize those information. See https://github.com/authgear/authgear-server/blob/2024-11-18.0/pkg/lib/infra/whatsapp/service.go#L86

In this PR, I centralized the handling of dev mode, test mode, as well as the audit log in messaging/Sender. But it is impossible for messaging/Sender to know the extra whatsapp information you mentioned. Those information is only available if Whatsapp is really configured.

Reply:
But I still think devMode should log any imformation that affects the resulting whatsapp message, just like sending is logging the whole email, but not the otp code & receiver only.

It is is difficult to do it in `messaging/Sender`, maybe we can create a method in `WhatsappSender` to do the logging?

> Those information is only available if Whatsapp is really configured.

I guess this is expected?

Reply:
OK. Let me separate the send operation into two steps.